org.eclipse.jdt.core/dom/org/eclipse/jdt/internal/core/dom/ICompilationUnitResolver.java

```diff
@@ -0,0 +1,113 @@
+import java.util.Map;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.WorkingCopyOwner;
+import org.eclipse.jdt.core.dom.ASTRequestor;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.FileASTRequestor;
+import org.eclipse.jdt.internal.compiler.batch.FileSystem.Classpath;
+
+/**
+ * This interface is used to resolve a jdt dom tree from source files.
+ * It is contributed to via the compilationUnitResolver extension point.
+ * This interface is currently internal only, and is not considered API.
+ * This interface may be modified, changed, or removed at any time.
+ *
+ * @since 3.38
+ */
+public interface ICompilationUnitResolver {
+	/**
+	 * Resolve the given source paths with the following options.
+	 *
+	 * @param sourceFilePaths the compilation units to create ASTs for
+	 * @param encodings the given encoding for the source units
+	 * @param bindingKeys the binding keys to create bindings for
+	 * @param requestor the AST requestor that collects abstract syntax trees and bindings
+	 * @param apiLevel Level of AST API desired.
+	 * @param compilerOptions Compiler options. Defaults to JavaCore.getOptions().
+	 * @param classpathList A list of classpaths to use during this operation
+	 * @param flags Flags to to be used during this operation
+	 * @param monitor A progress monitor
+	 */
+	void resolve(String[] sourceFilePaths, String[] encodings, String[] bindingKeys, FileASTRequestor requestor,
+			int apiLevel, Map<String, String> compilerOptions, List<Classpath> classpathList, int flags,
+			IProgressMonitor monitor);
+
+	/**
+	 * Parse the given source paths with the following options.
+	 *
+	 * @param compilationUnits the compilation units to create ASTs for
+	 * @param requestor the AST requestor that collects abstract syntax trees and bindings
+	 * @param apiLevel Level of AST API desired.
+	 * @param compilerOptions Compiler options. Defaults to JavaCore.getOptions().
+	 * @param flags Flags to to be used during this operation
+	 * @param monitor A progress monitor
+	 */
+	void parse(ICompilationUnit[] compilationUnits, ASTRequestor requestor, int apiLevel,
+			Map<String, String> compilerOptions, int flags, IProgressMonitor monitor);
+
+	/**
+	 *
+	 * @param sourceFilePaths the compilation units to create ASTs for
+	 * @param encodings the given encoding for the source units
+	 * @param requestor the AST requester that collects abstract syntax trees and bindings
+	 * @param apiLevel Level of AST API desired.
+	 * @param compilerOptions Compiler options. Defaults to JavaCore.getOptions().
+	 * @param flags Flags to to be used during this operation
+	 * @param monitor A progress monitor
+	 */
+	void parse(String[] sourceFilePaths, String[] encodings, FileASTRequestor requestor, int apiLevel,
+			Map<String, String> compilerOptions, int flags, IProgressMonitor monitor);
```

Comment:
in other consumers it was stated that Map<String, String> i snot

Reply:
I'm sorry, this comment is unclear.

Reply:
sry for the broken comment. 
In https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2497/files#diff-e70e3008348d77cb89e082387f8a534ee23302c5b4ad4243f114b917e8aee21dR68 it was stated that map<string,string> is not enough and you would need CompilerOptions

Reply:
Thanks for the update.  

That's a different PR with a different part of the code, specifically the compiler, and not the converter from source to dom. 

We prefer to pass around the raw settings, almost exclusively fetched from `JavaCore.getOptions()`.   However, the default implementation will then use that Map to create a CompilerOptions, and so should anyone making use of this non-API functionality. 

The CompilerOptions class is internal, and we prefer to not use it in this interface. If we have hopes that this interface may one day become API, we prefer to avoid (wherever possible) using classes that would then also need to be made public. 

If an implementation wishes to take our raw settings / Map, and use it to create a CompilerOptions class or their own implementation thereof, then we are ok with that.  But making CompilerOptions part of the interface doesn't seem right to us. 

Part of the issue here is that (from the comments) it appears they were downgrading FROM CompilerOptions TO Map, whereas we are loading the map from the raw settings and allowing any clients to convert to CompilerOptions at their leisure. 

I'm open to alternatives, but I'm also about to go on vacation, and considering the workflow of this particular part of the code, I genuinely believe using the raw map from settings in the interface is appropriate, and clients can expand that to the more fully-featured CompilerOptions as they find convenient to do so.

Reply:
The compiler options are indeed the right API at compiler level, but from the Java model (ICompilationUnit and so on), it looks like the Map is more commonly used, so it's consistent to stick to it.